heat/core/operations.py

```diff
@@ -731,11 +855,20 @@ def __reduce_op(x, partial_op, reduction_op, **kwargs):
             axis = (axis,)
         if isinstance(axis, tuple):
             partial = x._tensor__array
+            output_shape = x.gshape
             for dim in axis:
                 partial = partial_op(partial, dim=dim, keepdim=True)
-                shape_keepdim = x.gshape[:dim] + (1,) + x.gshape[dim + 1:]
-            shape_losedim = tuple(x.gshape[dim] for dim in range(len(x.gshape)) if not dim in axis)
-        output_shape = shape_keepdim if kwargs.get('keepdim') else shape_losedim
+                output_shape = output_shape[:dim] + (1,) + output_shape[dim + 1:]
+        if not keepdim:
+            gshape_losedim = tuple(x.gshape[dim] for dim in range(len(x.gshape)) if not dim in axis)
+            lshape_losedim = tuple(x.lshape[dim] for dim in range(len(x.lshape)) if not dim in axis)
+            output_shape = gshape_losedim
+            if reduction_op in [MPI_ARGMAX, MPI_ARGMIN]:
+                if 0 in axis:
+                    lshape_losedim = (2,) + lshape_losedim
+                else:
+                    lshape_losedim = (2 * lshape_losedim[0],) + lshape_losedim[1:]
```

Comment:
the 2 seems fishy to me. can you justify it for me? I am just not seeing it

Reply:
Hi Daniel @coquelin77, 
thanks, I'm really unhappy with this if statement. The problem is that, if axis is not None,  the result of local_argmin/max and MPI_ARGMIN/MAX is made up of two sets of results really: the min/max values along a dimension, and the respective indices. So at this stage the dimension along the reduced axis isn't 1 (like for every other reduction operation), it's 2. The min/max values get stripped out of the result at a later stage in the argmin/max() functions.

It's cumbersome though and giving me a lot of trouble pretty much whatever I'm working on. So @Markus-Goetz, @coquelin77  I'm thinking of  adding  __argreduce_op to operations.py and funnel argmin and argmax out to that function, leaving __reduce_op for "regular" reduction operations.

Reply:
OK here's what it looks like now (operations.py lines 217-221)
```
# Take care of special cases argmin and argmax: keep partial.shape[0]
            if (0 in axis and partial.shape[0] != 1):
                lshape_losedim = (partial.shape[0],) + lshape_losedim
            if (not 0 in axis and partial.shape[0] != x.lshape[0]):
                lshape_losedim = (partial.shape[0],) + lshape_losedim[1:]
```
Basically, the assumption is that whenever the first dimension of partial is different from what it should be (1 if reduction along axis 0, or x.lshape[0] if reduction along any other axis), there will be a good reason for it so just keep that partial[0] as first dimension.

This way we don't need to add ifs for every quirky reduction operation that comes our way.

Please let me know if I'm overseeing something.